Show Indels and Soft Clip alignments

danvk opened the pull request:
This also handles mismatches in the presence of more complex Cigar strings.

Fixes #85 
Fixes #111

![pileup indels](https://cloud.githubusercontent.com/assets/98301/7760494/2a309936-ffea-11e4-8550-f57616f0a21c.png)

<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/hammerlab/pileup.js/157)
<!-- Reviewable:end -->

ihodes commented on src/PileupTrack.js:
```diff
@@ -62,6 +61,65 @@ function makePath(scale, visualRead: VisualAlignment) {
 
+// Create the SVG element for a single Cigar op in an alignment.
+function enterSegment(parentNode, op, scale) {
+  if (op.op == 'M') {
```
Could be good to comment what 'M', 'D', etc are. I'm guessing match, deletion, insertion? Could also use const variables for them/remove magic strings.

danvk replied:
I factored out a `CigarOps` enum. It's much clearer now.

ihodes commented on src/PileupTrack.js:
```diff
@@ -222,7 +273,8 @@ class NonEmptyPileupTrack extends React.Component {
     var vReads = this.state.reads.map(
-        read => this.addRead(read, referenceSource));
+            read => this.addRead(read, referenceSource))
+        .filter(read => read.refLength);  // drop alignments w/o CIGARs
```
Would it be useful to warn or report on this?

danvk replied:
It's actually fairly common (see #111). I'm not sure what alignments without Cigar strings correspond to—maybe @arahuja or @smondet know more.

ihodes commented on src/pileuputils.js:
```diff
@@ -95,8 +86,95 @@ function getDifferingBasePairs(read: SamRead, reference: string): Array<BasePair
+      pos: refPos
+    });
+
+    switch (op.op) {
```
Same magic strings comment here and below.

ihodes replied:
(I realize it's CIGAR parsing, but it'd help to have it briefly documented here, anyway—your call though)

ihodes commented:
While not the most d3-y, (though maybe it is the *most* d3-y you can get in this situation?) this doesn't seem bad to me at all. 

In terms of making it a little more readable I'd *maybe* have gone for a case/switch on the op, and factor out the body of each to a function, but this is pretty manageable to read and understand as it is, with the 3 (+1) cases. 

Comments on what each branch is doing (e.g. just `// D: deletion, make a thin rectangle`) might be helpful coming back to this, but just having the 'D' be in a var like `DELETION_CIGAR_OP` would be more than sufficient, too.

Anyway, this all looks really fine, comments are all nits/minor suggestions.

danvk commented:
I factored out a `CigarOps` enum and changed to `switch`/`case`. I also moved the drawing code to all be in `updateSegment` and added comments indicating what's being drawn.
